# Transformers and BERT exploration

### 0. System and frameworks check.

In [8]:
import transformers

print(transformers.__version__)

5.17.0


### 1. BERT Tokenizer

In [1]:
# importing the essential 
from transformers import AutoTokenizer

# telling the transformers lib to load the best tokenizer for pre-trained bert model 
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

In [2]:
# starting with some small text 
text = "The patient has hypertension and chest pain."

tokens = tokenizer.tokenize(text)

print(tokens)

['the', 'patient', 'has', 'hyper', '##tension', 'and', 'chest', 'pain', '.']


The word "hypertension" is being splitted because of the BERT technique called WordPiece tokenization. 

Since that word isn't contained into its vocabulary, it divides it in two sub-parts. 

The "##" sign means that the piece who contains it has to be attached to the previous one.

In [ ]:
# going towards numerical translation of tokens
encoded = tokenizer(text, return_tensors="pt") # return as pytorch tensors instead of python list

print(encoded)

{'input_ids': tensor([[  101,  1996,  5776,  2038, 23760, 29048,  1998,  3108,  3255,  1012,
           102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


Where:

101 = [CLS] - start of sequence

102 = [SEP] - end of sequence

token_type_ids = to distinguish two different sequences

attention_mask = 1 / 0 boolean

with 1 = real token 

0 = padding

### 2. BERT Model

In [4]:
from transformers import AutoModel 

bert = AutoModel.from_pretrained("bert-base-uncased")

model.safetensors: reconstructing file:   0%|          |  0.00B /  440MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

[transformers] BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [10]:
outputs = bert(**encoded)

print(outputs.last_hidden_state.shape)

torch.Size([1, 11, 768])
